sql/core/src/main/scala/org/apache/spark/sql/execution/commands/parsers.scala

```diff
@@ -0,0 +1,420 @@
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.commands
+
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.spark.sql.{AnalysisException, SaveMode}
+import org.apache.spark.sql.catalyst.{CatalystQl, PlanParser, TableIdentifier}
+import org.apache.spark.sql.catalyst.analysis.UnresolvedRelation
+import org.apache.spark.sql.catalyst.expressions.{Ascending, Descending}
+import org.apache.spark.sql.catalyst.parser.{ASTNode, ParserConf, SimpleParserConf}
+import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, OneRowRelation}
+import org.apache.spark.sql.catalyst.plans.logical
+import org.apache.spark.sql.execution.commands._
+import org.apache.spark.sql.execution.datasources._
+import org.apache.spark.sql.types.StructType
+
+case class AlterTableCommandParser(base: CatalystQl) extends PlanParser {
+
+  def parsePartitionSpec(node: ASTNode): Option[Map[String, Option[String]]] = {
+    node match {
+      case Token("TOK_PARTSPEC", partitions) =>
+        val spec = partitions.map {
+          case Token("TOK_PARTVAL", ident :: constant :: Nil) =>
+            (unquoteString(cleanIdentifier(ident.text)),
+              Some(unquoteString(cleanIdentifier(constant.text))))
+          case Token("TOK_PARTVAL", ident :: Nil) =>
+            (unquoteString(cleanIdentifier(ident.text)), None)
+        }.toMap
+        Some(spec)
+      case _ => None
+    }
+  }
+
+  def extractTableProps(node: ASTNode): Map[String, Option[String]] = node match {
+    case Token("TOK_TABLEPROPERTIES", propsList) =>
+      propsList.flatMap {
+        case Token("TOK_TABLEPROPLIST", props) =>
+          props.map {
+            case Token("TOK_TABLEPROPERTY", key :: Token("TOK_NULL", Nil) :: Nil) =>
+              val k = unquoteString(cleanIdentifier(key.text))
+              (k, None)
+            case Token("TOK_TABLEPROPERTY", key :: value :: Nil) =>
+              val k = unquoteString(cleanIdentifier(key.text))
+              val v = unquoteString(cleanIdentifier(value.text))
+              (k, Some(v))
+          }
+      }.toMap
+  }
+
+  override def isDefinedAt(node: ASTNode): Boolean = node.text == "TOK_ALTERTABLE"
+
+  override def apply(v1: ASTNode): LogicalPlan = v1.children match {
+    case (tabName @ Token("TOK_TABNAME", _)) :: rest =>
+      val tableIdent: TableIdentifier = base.extractTableIdent(tabName)
+      val partitionSpec = base.getClauseOption("TOK_PARTSPEC", v1.children)
+      val partition = partitionSpec.flatMap(parsePartitionSpec)
+      matchAlterTableCommands(v1, rest, tableIdent, partition)
+    case _ =>
+      throw new NotImplementedError(v1.text)
+  }
+
+  def matchAlterTableCommands(
+      node: ASTNode,
+      nodes: Seq[ASTNode],
+      tableIdent: TableIdentifier,
+      partition: Option[Map[String, Option[String]]]): LogicalPlan = nodes match {
+    case rename @ Token("TOK_ALTERTABLE_RENAME", renameArgs) :: rest =>
```

Comment:
`rest` here is not alway Nil, it is possibly partition spec for alter table commands. This optional partition spec is after alter table command. Actually as we parse it before here, we don't need to use `rest` to catch it and can ignore it. I will update this.

For example,

    TOK_ALTERTABLE 2, 1, 22, 12 
    :- TOK_TABNAME 2, 5, 5, 12 
    :  +- table_name 2, 5, 5, 12 
    :- TOK_ALTERTABLE_COMPACT 3, 20, 22, 8 
    :  +- 'MAJOR' 3, 22, 22, 8 
    +- TOK_PARTSPEC 2, 7, 18, 34 
       :- TOK_PARTVAL 2, 10, 12, 34 
       :  :- dt 2, 10, 10, 34 
       :  +- '2008-08-08' 2, 12, 12, 37 
       +- TOK_PARTVAL 2, 15, 17, 51 
          :- country 2, 15, 15, 51 
          +- 'us' 2, 17, 17, 59